lib/aws-record/record/item_operations.rb

```diff
@@ -269,27 +269,35 @@ def _perform_save(opts)
             )
           end
         else
+          update_opts = {
+            table_name: self.class.table_name,
+            key: key_values
+          }
           update_pairs = _dirty_changes_for_update
           update_tuple = self.class.send(
             :_build_update_expression,
             update_pairs
           )
           if update_tuple
+            if opts.include?(:update_expression)
+              raise Errors::UpdateExpressionCollision,
+                    'Attributes need to be saved before custom update expression can be used.'
+            end
             uex, exp_attr_names, exp_attr_values = update_tuple
-            update_opts = {
-              table_name: self.class.table_name,
-              key: key_values,
-              update_expression: uex,
-              expression_attribute_names: exp_attr_names
-            }
-            update_opts[:expression_attribute_values] = exp_attr_values unless exp_attr_values.empty?
-          else
-            update_opts = {
-              table_name: self.class.table_name,
-              key: key_values
-            }
+            update_opts[:update_expression] = uex
+            # need to combine expression attribute names and values
+            update_opts[:expression_attribute_names] = [
+              exp_attr_names,
+              opts[:expression_attribute_names]
+            ].compact.reduce(&:merge)
```

Comment:
We can probably consolidate the names/values merging for both methods into 1 or 2 private methods. This repo likes to use private method calls with send, we could possibly re-use that code in transactions if it's not too awful.

Reply:
I pushed an additional commit to go ahead and share the logic. Let me know if this is in line with what you were thinking or if I should revert those changes.